src/jacobian.jl

```diff
@@ -166,11 +176,20 @@ end
 
 @inline function vector_mode_jacobian!(result, f::F, x::SArray{S,V,D,N}) where {F,S,V,D,N}
     ydual = vector_mode_dual_eval(f, x)
-    extract_jacobian!(result, ydual, N)
-    extract_value!(result, ydual)
+    result = extract_jacobian!(result, ydual, N)
+    result = extract_value!(result, ydual)
     return result
 end
 
+@inline function vector_mode_jacobian!(out::ImmutableDiffResult, f::F, x::SArray{S,V,D,N}) where {F,S,V,D,N}
+    ydual = vector_mode_dual_eval(f, x)
+    jout = extract_jacobian(ydual, x)
+    vout = extract_value(ydual, DiffBase.value(out))
+    out = DiffBase.jacobian!(out, jout)
+    out = DiffBase.value!(out, vout)
+    return out
+end
```

Comment:
I ran into two issues here.
1. `extract_jacobian!(result, ydual, N)` as written doesn't work because you cannot `setindex` on a reshaped `SArray`.
2. I think `extract_value!(result, ydual)` did work, but was slow.

Since the `@generated` function already existed for `extract_jacobian`, and it was straightforward to copy it for values, I used that approach here. I pass the value in `extract_value(ydual, DiffBase.value(out))` to get the shape right. In one of the tests the dimensions of `value(out)` are not the same as the native dimensions of `f(x)`.